datafusion/expr/src/udf.rs

```diff
@@ -17,21 +17,32 @@
 
 //! [`ScalarUDF`]: Scalar User Defined Functions
 
-use crate::{Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature};
+use crate::{
+    ColumnarValue, Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature,
+};
 use arrow::datatypes::DataType;
 use datafusion_common::Result;
+use std::any::Any;
 use std::fmt;
 use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
 /// Logical representation of a Scalar User Defined Function.
 ///
-/// A scalar function produces a single row output for each row of input.
+/// A scalar function produces a single row output for each row of input. This
+/// struct contains the information DataFusion needs to plan and invoke
+/// functions you supply such name, type signature, return type, and actual
+/// implementation.
 ///
-/// This struct contains the information DataFusion needs to plan and invoke
-/// functions such name, type signature, return type, and actual implementation.
 ///
+/// 1. For simple (less performant) use cases, use [`create_udf`] and [`simple_udf.rs`].
```

Comment:
> less performant

Hi, is there anyone who would like to explain a bit about why `create_udf()` is less performant than the UDFs created by `ScalarUDFImpl`?

Reply:
Filed https://github.com/apache/arrow-datafusion/pull/9384 to clarify docs

Reply:
Using `create_udf` create an extra indirection. Under the hood it's creating a 
```
pub struct SimpleScalarUDF {
    name: String,
    signature: Signature,
    return_type: DataType,
    fun: ScalarFunctionImplementation,
}

impl ScalarUDFImpl for SimpleScalarUDF {
    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
        (self.fun)(args)
    }
}
```

so it adds an extra call for every batch processed through the UDF

Reply:
That makes sense 👍  -- I don't think the overhead of a single function call is worth calling out in the docs however (I think it is more confusing than helpfl),  though please let me know if you disagree on https://github.com/apache/arrow-datafusion/pull/9384

Reply:
Nope. I agree it's not meaningful enough to call out tin docs